comit/src/network/protocols/orderbook.rs

```diff
@@ -1,22 +1,229 @@
+        );
+        tracing::info!("order published");
+        self.orders.insert(order.id, order.clone());
+
+        Ok(order.id)
+    }
+
+    /// Called by Alice i.e., the taker.  Does _not_ remove the order from the
+    /// order book.
+    pub fn take(&mut self, order_id: OrderId) -> anyhow::Result<()> {
+        // TODO: Refactor this
+        let peer_id = match self.peer_id_for_order(order_id) {
+            Some(id) => Ok(id),
+            None => Err(anyhow::Error::from(OrderbookError::OrderNotFound(order_id))),
+        }?;
+
+        self.take_order.send_request(&peer_id, order_id);
+
+        Ok(())
+    }
+
+    pub fn peer_id_for_order(&self, order_id: OrderId) -> Option<PeerId> {
+        if let Some(order) = self.orders.get(&order_id) {
+            Some(order.maker.peer_id())
+        } else {
+            None
+        }
+    }
+
+    /// Called by Bob i.e., the maker.
+    /// Does _not_ remove the order from the order book.
+    pub fn confirm(&mut self, peer_id: PeerId, _: OrderId, channel: ResponseChannel<Response>) {
+        let shared_swap_id = SharedSwapId::default();
+        tracing::info!(
+            "confirming take order request with swap id: {}",
+            shared_swap_id
+        );
+
+        self.take_order
+            .send_response(channel, Response::Confirmation(shared_swap_id));
+
+        // TODO: Do we want to do something with order_id? E.g. remove the order from
+        // the order book?
+
+        // TODO: Work around the fact that libp2p silently drops the channel send error
+        // and only push this event if we know the response got to peer.
+        self.events
+            .push_back(BehaviourOutEvent::TakeOrderConfirmation {
+                peer_id,
+                shared_swap_id,
+            });
+    }
+
+    /// Called by Bob i.e., the maker to deny a take order request.
+    pub fn deny(&mut self, channel: ResponseChannel<Response>) {
+        self.take_order.send_response(channel, Response::Error);
+    }
+
+    /// Removes the order from the local order book i.e., does nothing on the
+    /// network.
+    pub fn local_remove_order() -> Option<OrderId> {
+        unimplemented!()
+    }
+
+    pub fn get_orders(&self) -> Vec<Order> {
+        #[allow(clippy::map_clone)]
+        self.orders.values().map(|order| order.clone()).collect()
+    }
+
+    pub fn get_order(&self, order_id: &OrderId) -> Option<Order> {
+        if let Some(order) = self.orders.get(order_id) {
+            Some(order.clone())
+        } else {
+            None
+        }
+    }
+
+    pub fn get_trading_pairs(&self) -> Vec<TradingPairTopic> {
+        let mut topics = vec![];
+        for topic in self.trading_pairs.iter() {
+            topics.push(topic.clone());
+        }
+        topics
+    }
+
+    pub fn add_trading_pair(&mut self, topic: &TradingPairTopic) {
+        self.trading_pairs.insert(topic.clone());
+    }
```

Comment:
Generally, arguments should take whatever you need. If you need to own `topic`, then don't make it a reference.

If the caller has an owned variant and no longer needs it after your call, we are wasting an allocation by cloning it here regardless :)